auth/auth.go

```diff
@@ -35,79 +74,211 @@ func New(authURL string, insecure bool) (client *Client, err error) {
-	LastLogin    string `json:"last_login,omitempty"`
-}
+// Login to Quarterdeck, storing the API credentials on the client and making a login
+// request to Quarterdeck to fetch access and refresh tokens. Ensure that a context
+// with a deadline is specified in order to reduce how long the client attempts to login
+// for. Once logged in, the authentication client can hand out credentials on demand.
+func (c *Client) Login(ctx context.Context, clientID, clientSecret string) (creds credentials.PerRPCCredentials, err error) {
+	// Require both clientID and clientSecret
+	if clientID == "" || clientSecret == "" {
+		return nil, ErrIncompleteCreds
+	}
 
-func (c *Client) Login(ctx context.Context, clientID, clientSecret string) (_ credentials.PerRPCCredentials, err error) {
+	// Store the API key on the client so that authentication can happen again.
 	c.apikey = &APIKey{
 		ClientID:     clientID,
 		ClientSecret: clientSecret,
 	}
 
-	if err = c.Authenticate(ctx); err != nil {
+	// Authenticate and store the tokens on the client to cache for each call.
+	if c.tokens, err = c.Authenticate(ctx, c.apikey); err != nil {
+		return nil, err
+	}
+
+	// Return credentials for dial options.
+	return c.Credentials(ctx)
+}
+
+// Credentials returns the PerRPC credentials to make a gRPC request. If the tokens are
+// expired, this method will refresh them by making a request to Quarterdeck. An error
+// is returned if the client is not logged in. This method should be called before every
+// Ensign RPC in order to ensure the RPC has valid credentials.
+func (c *Client) Credentials(ctx context.Context) (_ credentials.PerRPCCredentials, err error) {
+	// Check if tokens exist; if they don't exist, then authenticate.
+	if c.tokens == nil || c.tokens.AccessToken == "" || c.tokens.RefreshToken == "" {
+		// Tokens are missing or are partial, authenticate to get new tokens
+		if c.tokens, err = c.Authenticate(ctx, c.apikey); err != nil {
+			return nil, err
+		}
+	}
+
+	// Check if the access token is valid
+	var accessValid bool
+	if accessValid, err = c.tokens.AccessValid(); err != nil {
+		// Returning an error here is acceptable because we checked if the access tokens
+		// were missing in an above step.
 		return nil, err
 	}
 
-	creds := &Credentials{
+	// If the access token is not valid, attempt to use the refresh token to validate.
+	if !accessValid {
+		var refreshValid bool
+		if refreshValid, err = c.tokens.RefreshValid(); err != nil {
+			// Returning an error is acceptable here because we checked if the refresh
+			// tokens were missing in an above step.
+			return nil, err
+		}
+
+		// If the refresh tokens are valid, use it to refresh the access token,
+		// otherwise reauthenticate using the credentials.
+		if refreshValid {
+			if c.tokens, err = c.Refresh(ctx, c.tokens); err != nil {
+				return nil, err
+			}
+		} else {
+			if c.tokens, err = c.Authenticate(ctx, c.apikey); err != nil {
+				return nil, err
+			}
+		}
```

Comment:
Right now the methodology is to wait until an request to this method comes in to check if the credentials are valid or not and if they're not to make a call to Quarterdeck. We could also start a background routine to refresh the tokens when they expire; should I make a story to do this? Perhaps it could just be optional?

Reply:
Maybe it's something we should talk about during office hours then.